ntp-udp/src/lib.rs

```diff
@@ -47,18 +49,22 @@ impl UdpSocket {
         peer_addr = debug(self.as_ref().peer_addr()),
         buf_size = buf.len(),
     ))]
-    pub async fn send(&self, buf: &[u8]) -> io::Result<usize> {
+    pub async fn send(&self, buf: &[u8]) -> io::Result<(usize, Option<NtpTimestamp>)> {
         trace!(size = buf.len(), "sending bytes");
         loop {
             let mut guard = self.io.writable().await?;
             match guard.try_io(|inner| inner.get_ref().send(buf)) {
-                Ok(result) => {
-                    match &result {
-                        Ok(size) => trace!(sent = size, "sent bytes"),
-                        Err(e) => debug!(error = debug(e), "error sending data"),
+                Ok(result) => match result {
+                    Err(e) => {
+                        debug!(error = debug(&e), "error sending data");
+                        return Err(e);
                     }
-                    return result;
-                }
+                    Ok(size) => {
+                        trace!(sent = size, "sent bytes");
+                        let send_timestamp = fetch_send_timestamp(self.io.get_ref())?;
```

Comment:
I feel like it would be better to be a little more careful when retrieving the send timestamp. I'm unsure about the blocking behavior here (what if the kernel/driver responds too late and the recvmsg call is too early), I would prefer not to error on EWOULDBLOCK but instead have the same behavior as with our other recvmsg call. Secondly there could be other errors in the error queue, I think it would be best to keep to the timestamping docs (https://www.kernel.org/doc/Documentation/networking/timestamping.txt) and look for a message with the correct control message set. I'm unsure what to do about any other errors (or what could even appear here), I feel that ignoring other messages here is fine since we weren't looking at them before. But better to be safe here, or if we don't need that make sure we really document that it isn't needed.

Reply:
> I would prefer not to error on EWOULDBLOCK but instead have the same behavior as with our other recvmsg call

done

> Secondly there could be other errors in the error queue, I think it would be best to keep to the timestamping docs (https://www.kernel.org/doc/Documentation/networking/timestamping.txt, see 2.1.1) and look for a message with the correct control message set. 

so that would mean continue to receive until we find a message with the timestamping control message? and skip (but report) any other messages in that queue?